many: fix travis; fix native and cross-compiling

anonymouse64 opened the pull request:
This PR fixes cross-compiling the gadget snap, which now works properly everywhere, on LP, locally with lxd or multipass on native arches, or also cross-compiled with lxd/multipass, etc. Snapcraft 4.0 no longer supports --target-arch, so we no longer use that and this snapcraft.yaml now correctly builds either from amd64 to armhf or from armhf to armhf with supported snapcraft features in the architectures key of the snapcraft.yaml.

Building locally from this will also work if the user added the snappy-dev image PPA manually with --destructive-mode too.

anonymouse64 committed:
snapcraft.yaml, sources.list.cross: use the snappy-dev PPA

This will fix cross-compiling to use the right version of the flash-kernel, etc.
packages.

We need snapcraft 4.0 to use the package-repositories key in the snapcraft.yaml
schema, which also means we need to use the experimental flag when building to
enable it.

Signed-off-by: Ian Johnson <ian.johnson@canonical.com>

anonymouse64 committed:
snapcraft.yaml: fix some inconsistent white-space

Signed-off-by: Ian Johnson <ian.johnson@canonical.com>

anonymouse64 committed:
snapcraft.yaml: don't use SNAPCRAFT_PROJECT_DIR for psplash part local tree

The more modern thing to do here that works everywhere snapcraft works is to put
the local parts we need into the stage dir, then don't prime anything from the
stage dir, but use the stage dir from the part that we need.

Signed-off-by: Ian Johnson <ian.johnson@canonical.com>

anonymouse64 committed:
snapcraft.yaml: add missing build-deps

These are installed by default on the LP builders, but not for example in lxd
containers when building locally.

Signed-off-by: Ian Johnson <ian.johnson@canonical.com>

anonymouse64 committed:
snapcraft.yaml: support cross-compiling w/o --target-arch

Compiling with --target-arch is no longer supported by snapcraft, so instead
because this branch will only ever build for armhf, we can detect if we are
doing a native or cross-compilation build and just set the arch triplet
ourselves accordingly.

This also means we need to switch from using snapcraft's on: to: advanced
grammar for build-packages to just using try: which will work both natively and
when cross-compiling because failures trying to install these packages natively
will just get ignored.

Signed-off-by: Ian Johnson <ian.johnson@canonical.com>

anonymouse64 committed:
Makefile, snapcraft.yaml: specify Dir::Etc::sourceparts too

In recent focal builder images that one would use to build this snap locally,
when cross-compiling there are now files in this directory in the build
environment which we want to ignore when downloading manually with our
stage-package helper in the Makefile. However, when natively compiling we will
want to copy these sources from the host in case the snappy-dev image PPA was
installed on the host, so that we can still build using the correct sources
here.

Signed-off-by: Ian Johnson <ian.johnson@canonical.com>

anonymouse64 committed:
snapcraft.yaml: use SNAPCRAFT_PART_SRC and source: . for gadget part

This enables proper building with remote-build, otherwise the part has no source
and doesn't effectively get copied and doesn't build when it otherwise would
build.

Signed-off-by: Ian Johnson <ian.johnson@canonical.com>

anonymouse64 committed:
sources.list.cross: use focal instead of bionic

Signed-off-by: Ian Johnson <ian.johnson@canonical.com>

anonymouse64 committed:
.travis.yml: build the snap via lxd

snapcraft 4.0 enables us to build core20 images with lxd, meaning we can
properly test the snap out. Travis doesn't yet support focal, so building the
snap on destructively on bionic doesn't correctly test anything.

snapcraft 4.0 also doesn't support building as root, so instead we build as the
user, but ensure that the CI user is in the lxd group (because it is not by 
default it seems).

Finally, note that this results in a cross-compile build, which is also 
currently broken because it will not use the snappy-dev PPA, but it will at 
least use the correct arch packages. We will fix it to use the snappy-dev PPA w/
a future commit.

Signed-off-by: Ian Johnson <ian.johnson@canonical.com>

anonymouse64 commented:
I will port this to the 20-arm64 branch later.

xnox reviewed (changes requested):
Why do current snappy-dev PPA arm64 builds work, and armhf ones don't?

What is the smallest diff / change to unbreak that, as a snap build in launchpad? Note launchpad does setup etc/apt/sources and uses destructive mode to build snaps.

Have you created a snap in launchpad with these changes, and can you point at build logs?

These changes are a lot more than what I was expecting.

anonymouse64 commented:
> Have you created a snap in launchpad with these changes, and can you point at build logs?

Yes, see https://launchpad.net/~anonymouse67/+snap/pi-gadget-snap and specifically build log: https://launchpad.net/~anonymouse67/+snap/pi-gadget-snap/+build/930577 for armhf native build on LP. 

For the other use cases, here's the log from a focal install on an armhf device where the user manually adds the snappy-dev PPA: https://pastebin.ubuntu.com/p/xZT6fMXXmM/.

anonymouse64 commented:
Also note that this PR will fix the travis CI checks to the extent that they are actually checking that the builds work on focal rather than bionic as they were previously. The travis job is still somewhat wrong because it's not using the snappy-dev PPA, but that will be fixed in the follow-up that is currently blocked on waiting for snapcraft 4.0 to be fixed as per that PR.

anonymouse64 committed:
sources.list.cross: rename to cross-sources.list

This eliminates warning when cross-compiling such as

N: Ignoring file 'sources.list.cross' in directory '/home/user/git/pi-gadget/parts/gadget/src/stage/apt/multiverse.sources.list.d/' as it has an invalid filename extension

which happens because in the cross-compiling case we copy files from the helpers
and the filename isn't recognized by apt as a valid source file. The message is
harmless, but not ideal when looking through logs as it can be confusing.

Signed-off-by: Ian Johnson <ian.johnson@canonical.com>

anonymouse64 commented:
I decided to update this to just use the package-repositories key to build the snap and not do that in a different PR, so this now properly does cross-compilation too.

It also fixes travis to work with snapcraft 4.0 because with 4.0 you can't run snapcraft with sudo.

anonymouse64 commented:
For lp build example see https://launchpad.net/~anonymouse67/+snap/pi-gadget-snap/+build/992153

anonymouse64 commented:
gentle ping @xnox and @waveform80

knitzsche commented:
I just built a gadget from https://github.com/anonymouse64/pi-gadget/tree/bugfix/uc20-backport-arm64-to-armhf and it seemd to work fine in a uc20 pi image. Got through console-conf to shell on boot.
I like the conservation of cross compile (amd64 - > armhf in this case) so that I can build the gadget on amd64 workstation.

xnox commented on snapcraft.yaml:
```diff
@@ -80,13 +109,9 @@ parts:
-
+      - try:
+          - gcc-aarch64-linux-gnu:amd64
+          - libc6-dev-arm64-cross:amd64
```
Is this branch targetting only arm64, or only armhf, or both?

Cause above triplet is set to armhf, yet only arm64 cross-compiler is installed. Unless things that are build here need compilation for armhf tripplet irrespective of armhf/arm64 (freestanding?! but then it would not run on a subset of AArch64 cpus without the optional AArch32 instruction set)

anonymouse64 replied:
ah thanks for catching this, this is actually a problem, since this PR is targeting the armhf branch, this should be

```suggestion
          - libc6-dev-armhf-cross:amd64
```
(or whatever the actual packages are called)

but to answer your question, the `try` is always executed on all architectures, but it's not treated as fatal if snapcraft doesn't find this package available when it is building. This is because the `on ... to ...` support in snapcraft is broken because --target-arch is what sets the arch in `to ...` but we can't set that anymore except when building dangerously.

anonymouse64 replied:
fixed this

xnox commented on snapcraft.yaml:
```diff
@@ -2,27 +2,41 @@ name: pi
+    source: .
     override-build: |
+      set -x
+      if [ "$SNAPCRAFT_ARCH_TRIPLET" = "x86_64-linux-gnu" ] || [ -z "$SNAPCRAFT_ARCH_TRIPLET" ]; then
```
Why is this if block needed? Surely both in native, and cross case, BUILD_ARCH_TRIPLET must have the same value, no?

anonymouse64 replied:
fair we probably can just unconditionally set BUILD_ARCH_TRIPLET like this, I will test this out

anonymouse64 replied:
this works to set it unconditionally

xnox commented on snapcraft.yaml:
```diff
@@ -2,27 +2,41 @@ name: pi
       if [ "$(arch)" != "aarch64" -a "$(arch)" != "armv7l" ]; then
         echo "Cross compilation detected; using pre-defined sources list"
-        OPTIONAL_ARGS="SOURCES_HOST=\"./helpers/sources.list.cross\""
+        OPTIONAL_ARGS="SOURCES_HOST=\"./helpers/cross-sources.list\" SOURCES_D_HOST=\"./helpers\""
```
The sources.list is arch qualified, and is correct for either native or cross compilation, does it need to be in a conditional?

anonymouse64 replied:
yes you're right it can always be used

tobias-grasse commented:
I've been using this in my fork of this gadget snap, and have had no issues so far. Tested with multiple Raspberry Pi models: 3B, 3B+, 4B. Are there any issues that speak against merging this? Upstream integration is always better than patches in git stashes 😄 Thanks in advance!

anonymouse64 commented:
@tobias-grasse sorry this is blocked waiting for me to address @xnox's feedback, I will do so soon hopefully now that snapd 2.48 is mostly out the door

anonymouse64 committed:
snapcraft.yaml: install the right cross-compiling toolchain

Signed-off-by: Ian Johnson <ian.johnson@canonical.com>

anonymouse64 committed:
snapcraft.yaml: always set BUILD_ARCH_TRIPLET and OPTIONAL_ARGS

The cross-compiling sources.list is valid for all architecture situations, and
the arch-triplet is always the same, regardless of whether we are
cross-compiling or not.

Signed-off-by: Ian Johnson <ian.johnson@canonical.com>

anonymouse64 commented:
Alright so this branch has been updated to address @xnox's feedback, I force-pushed it to rebase on top of 20-armhf upstream here. I tested building it with the following methods:
* [x] with an amd64 Launchpad snap recipe that builds out of the snappy-dev/image PPA, see this link for an example build: https://code.launchpad.net/~anonymouse67/+snap/pi-gadget-snap/+build/1255460 (but see below for the caveat)
* [x] locally on an amd64 system using LXD provider for snapcraft (thus cross-compiling)
* [x] natively on a raspberry pi 4 running Ubuntu Server 20.04

Thus, this should be ready now, but note that we should probably hold off on merging it for just a little bit more, since in order to cross-compile this gadget properly with i.e. LXD or multipass and vanilla snapcraft, you still need to pass `--enable-experimental-package-repositories`, but that makes the snap not buildable on LP, as LP doesn't yet know how to handle additional snapcraft arguments when trying to build a git repo / snap recipe. 

What we could do is I could revert that patch adding `package-repositories` and we merge this as-is, and open a followup just adding package-repositories which we don't merge until snapcraft has landed that out of experimental. I spoke with the snapcraft team today about package-repositories and they are apparently very close to getting that out of experimental for the next release, which is due at the end of this week, so maybe just waiting for this PR is fine.

In any case, I would like to get this PR +1d again before porting the relevant patches to the arm64 branch.
